src/workerd/io/container-client.c++

```diff
@@ -0,0 +1,334 @@
+    // fails because the app hasn't finished starting up or because the mapping is outdated.
+    // To be safe we should inspect the container to get up to date mappings.
+    auto portMappings = kj::get<1>(co_await containerClient.inspectContainer());
+    auto maybeMappedPort = portMappings.find(containerPort);
+    if (maybeMappedPort == kj::none) {
+      throw KJ_EXCEPTION(DISCONNECTED, "connect(): Connection refused: container port not found");
+    }
+    auto mappedPort = KJ_ASSERT_NONNULL(maybeMappedPort);
+
+    auto request = containerClient.network->connect(
+        kj::str(containerHost, ":", mappedPort), headers, kj::HttpConnectSettings());
+
+    auto status = co_await request.status;
+    if (status.statusCode != 200) {
+      throw KJ_EXCEPTION(DISCONNECTED, "connect(): Connection refused");
+    }
+
+    auto upPipe = kj::newOneWayPipe();
+    auto upEnd = kj::mv(upPipe.in);
+    auto results = context.getResults();
+    results.setUp(containerClient.byteStreamFactory.kjToCapnp(kj::mv(upPipe.out)));
+    auto downEnd = containerClient.byteStreamFactory.capnpToKj(context.getParams().getDown());
+    pumpTask = kj::joinPromisesFailFast(
+        kj::arr(upEnd->pumpTo(*request.connection), request.connection->pumpTo(*downEnd)))
+                   .ignoreResult()
+                   .attach(kj::mv(upEnd), kj::mv(request.connection), kj::mv(downEnd));
+    co_return;
+  }
+
+ private:
+  ContainerClient& containerClient;
+  kj::String containerHost;
+  uint16_t containerPort;
+  kj::Maybe<kj::Promise<void>> pumpTask;
+};
+
+kj::Promise<ContainerClient::Response> ContainerClient::dockerApiRequest(
+    kj::HttpMethod method, kj::StringPtr endpoint, kj::Maybe<kj::StringPtr> body) {
+  kj::HttpHeaderTable::Builder builder;
+  const auto userAgent = builder.add("User-Agent");
+  const auto accept = builder.add("Accept");
+  const auto contentType = builder.add("Content-Type");
+  const auto contentLength = builder.add("Content-Length");
+  auto headerTable = builder.build();
+  kj::HttpHeaders headers(*headerTable);
+  headers.set(kj::HttpHeaderId::HOST, "localhost");
+  headers.set(userAgent, "workerd/1.0");
+  headers.set(accept, "*/*");
+
+  KJ_IF_SOME(requestBody, body) {
+    headers.set(contentType, "application/json");
+    headers.set(contentLength, kj::str(requestBody.size()));
+
+    auto req = httpClient->request(method, endpoint, headers, requestBody.size());
+    {
+      auto body = kj::mv(req.body);
+      co_await body->write(requestBody.asBytes());
+    }
+    auto response = co_await req.response;
+    auto result = co_await response.body->readAllText();
+    co_return Response{.statusCode = response.statusCode, .body = kj::mv(result)};
+  } else {
+    auto req = httpClient->request(method, endpoint, headers);
+    { auto body = kj::mv(req.body); }
+    auto response = co_await req.response;
+    auto result = co_await response.body->readAllText();
```

Comment:
Is the amount of text returned here determined by us or users? If it is determined by users this could be problematic for memory management.

Reply:
Determined by the docker api. I don't really think we should worry too much about memory management here.
If the user's docker service is returning a response that's bigger than their computer's memory budget it's really their problem.